add searching and sorting

PiotrekPaprocki opened the pull request:
Hi,

I've made a few changes to px-dropdown needed in app I'm developing on predix.
Some of the changes are similar to what @budsalvetti did in his #19  pull request.

 ## A description of the changes proposed in the pull request:

- added an option to search thorough the items
- added and option to sort the items. Works only with `checkboxMode='true'`. Useful especially when there is a lot of items to pick from
- added displaying all selected values separated by coma in the header of dropdown instead of word "Select"

 ## @mentions of the person or team responsible for reviewing proposed changes:

@runn-vermel @benoitjchevalier Could you please look at it? :)


Also when developing, app is using files from bower_component/px-dropdown instead of those in base folder, so every time I want to check changes I have to update bower_component/px-dropdown. It should be like that?

EDIT: I see one test is failing, is there any guide about how to run tests locally?

sumeet-singh04 commented:
> Also when developing, app is using files from bower_component/px-dropdown instead of those in base folder, so every time I want to check changes I have to update bower_component/px-dropdown. 

This happened to me also. The component has a dependency to px-demo which in turn is dependent to px-dropdown. This creates a circular dependency and the original px-dropdown.html or any other file doesn't gets served.

randyaskin commented:
There are instructions here for using WCT for local testing: https://github.com/Polymer/web-component-tester#running-your-tests

runn-vermel commented:
@sumeet-singh04 

> This happened to me also. The component has a dependency to px-demo which in turn is dependent to px-dropdown. This creates a circular dependency and the original px-dropdown.html or any other file doesn't gets served.

The component has a DEV dependency on px-demo, not a dependency - there is no circular dependency issue here. 

Ideally, you'd never be editing our components - if you do, you risk them getting overwritten - @PiotrekPaprocki did the right thing by doing a PR, but until the PR is accepted, there's no way around this in production. on your local machine, you can always use bower link to create a link to your modified version - which you would put outside of the bower_components folder - and then link to that inside you app until the PR is accepted.

sumeet-singh04 commented:
@runn-vermel  I am talking about developing on px-dropdown not using px-dropdown inside another project. So when you do a local setup of px-dropdown and do a bower install, it pulls in px-dropdown inside bower_components as it is a dev dependency to px-demo. 

Now if you run the gulp serve, to start a local server, it creates route to the root of the project as well as to bower_components.

The problem occurs here, as my changes are not reflecting in the local demo as it's reading px-dropdown.html that is present inside the bower_components rather than the one present at the root of the project.

runn-vermel commented:
@sumeet-singh04 - thanks for the explanation - that makes sense. have you tried doing using bower link to get around this issue?

sumeet-singh04 commented:
@runn-vermel No, not yet. But I guess that should work and fix it. Thanks for suggesting this. I will try that.

PiotrekPaprocki commented:
@randyaskin I spent few hours on trying to first simply install wct globally (thx corp proxy) and then to connect to selenium but I gave up at least for now. 
So I've made few changes based on output from travis ci but I hit the same case :< 
Could you tell me if travis uses tests from master or pull requests? I couldn't info about that.

PiotrekPaprocki commented:
I used my personal pc to fix the issue. I wanted to stay consistent inside of the px-dropdown-content (ie. leave `<li>` tags for search `input`) and I've made slight changes to the failing test (I don't know why `px_dropdown_content.$.dropdown.querySelectorAll('input[type=checkbox]');` in line 261 grabs the search input too). I hope it's fine. Sorry for a little mess with previous commit.

runn-vermel commented on px-dropdown-content.html:
```diff
@@ -299,46 +337,61 @@
             this._checkboxChanged(checkbox);
           }
+
+          this.items.forEach(function (item) {
```
we try really hard not to use forEach because it's so much slower - can you please re-write this with a regular for loop?

PiotrekPaprocki replied:
I was following solutions already used in the component:). I can re-write also two other forEach loops if you wish.

runn-vermel replied:
That would be great :)

PiotrekPaprocki replied:
Done and tested :)

runn-vermel commented on px-dropdown-content.html:
```diff
@@ -407,6 +460,52 @@
+      } else {
+        // return a filter function for the current search string
+        searchTerm = searchTerm.toLowerCase();
+        return function(entry) {
```
i'm not clear why this return is inside a function?

PiotrekPaprocki replied:
`filter` property of `dom-repeat` template needs function in return statement to work properly with dynamic values (like when we want to use some input search with custom values). 
I was following [Polymer docs](https://www.polymer-project.org/1.0/docs/devguide/templates#dynamic-sort-and-filter-changes)

runn-vermel replied:
Had to look into this - now it makes perfect sense :) 👍

PiotrekPaprocki commented on px-dropdown-content.html:
```diff
@@ -407,6 +460,52 @@
+      }
+    },
+    /**
+     * This function sort the items
```
I hope it's more understandable now (and not too long/complicated).

PiotrekPaprocki commented on sass/px-dropdown-content-sketch.scss:
```diff
@@ -52,6 +52,20 @@ Predix branding rules go in px-dropdown-predix.scss, not in this file.
 
+
+input[type="search"] {
+  margin-left: 8px;
```
Does the way I did it is ok, or you meant something else, like add class directly to the html element?

runn-vermel replied:
sorry, yes, please add the correct class ( i think it'll b u-ml++)

PiotrekPaprocki replied:
done

runn-vermel commented:
I might be missing something, but, looking at the demo in the PR, and clicking on the checkbox search mode, I don't see any difference between that and the checkbox mode?

PiotrekPaprocki commented:
I mentioned it in the first comment ( also sumeet-singh04 elaborated more specific about it [click](https://github.com/PredixDev/px-dropdown/pull/32#issuecomment-279511462) ). Basically px-dropdown-demo component is using bower_component/px-dropdown instead of px-dropdown from the root directory, so if you want to see your changes you have to copy html and css files to bower_component/px-dropdown.
It is strange behavior for dev environment and your astonishment seems to confirm it :)

runn-vermel commented:
@PiotrekPaprocki 

Ah, I see! to achieve what you want, type `bower link` in the component itself, and then type `bower link px-dropdown`. you might get an error on the screen, but if you look in your `bower_components` folder, you'll see the simlink.

Having said that, I was finally able to actually VIEW the Pull Request, and i'm sending it to our designers - looks like you're almost there, but you're missing spacing, and you're not using our input boxes from px-forms-design. I'll post their reply here once I get it.

PiotrekPaprocki commented:
Awesome, can't wait :)

PiotrekPaprocki commented:
@runn-vermel Any update? :)

runn-vermel commented:
@randyaskin - has design looked at this yet?

randyaskin commented:
Working on merging this and updating the design a bit. First, I switched to class `text-input input--search` from px-forms-design to standardize the appearance of the search field:
![screen shot 2017-03-07 at 12 40 59 pm](https://cloud.githubusercontent.com/assets/2130613/23677025/709fccea-0333-11e7-8a21-b8ac09ea995e.png)

Also found some weirdness regarding the behavior. If you enable search-mode, but not checkbox-mode, then search for an item further down the list and click on it, the selected value actually ends up being something *else* from the list - something weird is going on between the dom-repeat filter and the click event target.

For the comma separated list of selected values, they seem to populate in a random order - I would expect them to appear sequentially based on the order I click them. When the items are all unchecked, it would be nice if the placeholder text would come back in.

And finally, when sort-mode is enabled, there's some odd behavior as well - checking and unchecking items causes random rearrangement of other items in the list. Would it be easy to also sort alphabetically (especially when not in checkbox mode)?

Can you take a look at some of these, then I'll take another stab at merging? Thanks!

PiotrekPaprocki commented:
> Also found some weirdness regarding the behavior. If you enable search-mode, but not checkbox-mode, then search for an item further down the list and click on it, the selected value actually ends up being something else from the list - something weird is going on between the dom-repeat filter and the click event target.

Reason: `px-dropdown-value-changed` fired after click on item grabs values from property `computedItems` which is not update after change searchTerm. Fixed. Pushed. 


> For the comma separated list of selected values, they seem to populate in a random order - I would expect them to appear sequentially based on the order I click them. When the items are all unchecked, it would be nice if the placeholder text would come back in.

I'm pretty sure they populate in their keys order. Items with lower key goes first, so if you pick items with key 1 and 3 and after that item with key=2, the last item goes between key=1 and key=3 item. 

Also items seems to be displayed in a bit random order from the very beginning, even at the prntscreen you provided. First "Rhino" has key=11, second "Rhino" key=1, some "Giraffe" in the middle has key=2 etc. Maybe those animals behave like that in real world but I think it's a bit odd for the component :)
ComputedItems property has items in right order so maybe it's sth with dom-repeat.

I removed placeholder intentionally, because I thought it can be misleading - user could not be sure if he picked "Selected" value from the list or it's just a placeholder (or make placeholder text more transparent/darker/sth). It's empty by default in Checkbox Search Mode, but you will see the last value it had in previous mode - demo component doesn't reset this. 

> And finally, when sort-mode is enabled, there's some odd behavior as well - checking and unchecking items causes random rearrangement of other items in the list. Would it be easy to also sort alphabetically (especially when not in checkbox mode)?

Indeed, it's pretty crazy how they behave right now. 
I'm thinking of sort them in three groups:
- disabled items at the end
- checked first
- the rest in the middle

Also all of them sort alphabetically in their specific groups. 

Do you want me to set alphabetical sort by default everywhere, or maybe add a property like `alphabetical`?

If you didn't fall asleep reading yet please let me know what do you think of my ideas (especially second one, do you want it as you said or we're going a bit in my way:) ) and the commit.

randyaskin commented:
All of that sounds good. Probably best to make a property / option for sorting by value alphabetically vs sorting by key - I could see some use cases where you might want the options in a specific order.

PiotrekPaprocki commented:
Changes:
Now sortMode is a String and specify sorting mode:
`sort-mode='key'` - sorting by key,
`sort-mode='val'` - sorting by value
Default value: `key`

Function `_computeSort` is now quite ugly. But my goal was to cover all possibilities: 
- split items in three groups, 
- disable items can also be checked so I put them at the end of checked group (test it if you wish, just add `checked: true` to some disable items);
- two modes of sorting

and all of that using Polymer's template sort property which is based on `Array.sort()`. Cleaner solution would be to use i.e. `map()` first to split items in three groups and then `sort()` on each group, but this involve manipulation on whole items list not only two values used in `sort()`. 

Items in coma-separated list of selected values are ordered like in the dropdown content.  

Is it allowed of using es6 syntax, like arrow functions? (I used labda once). 

How about placeholder, do we leave it empty?

randyaskin commented:
I've started merging it and updating the styles in the "search" branch [here](https://github.com/PredixDev/px-dropdown/tree/search). Arrow functions are going to be problematic, as we have to support IE11 still - some of our newer components are set up to use Babel for compiling IE-compatible code but dropdown isn't set up for that yet.

I still think it would be cleaner to put the placeholder text back in when all the checkboxes are unchecked.

Thanks again, this is looking great!

PiotrekPaprocki commented:
I did get rid of arrow function and put placeholder "Select" back.

Also I merged branch search with my fork, so there is less to merge for you after my changes(but one dirty commit more:<)

randyaskin commented:
I love how unit tests that have always passed before and have nothing to do with your changes are suddenly failing... silly WCT 😝

PiotrekPaprocki commented:
Silly me ;) I should run the test, but still I have some problems with wct when behind corp proxy... My bad, I'll check it tomorrow.

randyaskin commented:
The tooltip is showing up just fine for long list items in the demo page, but for some reason it isn't showing up in the test fixture page...

PiotrekPaprocki commented:
My quick review:

Tests starts failing from `6c8915b594504c54db5c24cc0000309240342483`.

`check if in checkbox mode we can toggle check state` test stopped failing after remove default value of `sortMode`, so `_initSort` wasn't invoked. 

In line 95 of `px-dropdown-custom-tests.js` if we console.log li.textContent before and after the commit it looks differently - supposed reason of first test failing. If we remove observer from sortMode first test pass.

Conclusion:
The reason behind those tests failing is rerendering `dropdownItems` template.

That's it for today, I'm going to bed. I'm gonna play with I hope tomorrow if I'll find some time, if not - later this week.

PiotrekPaprocki commented:
`<my-thoughts>` 
Calling `render()` forces polymer template to rerender itself even if some of the parameters weren't passed yet, i.e. `maxContCharacterWidth`. So we call `render()` on template before it's "ready", template appears, test runs but without `maxContCharacterWidth` which wasn't injected yet. 
`</my-thoughts>`

Adding `this.async` to `_initSort` solves the problem, but it's changing the code to pass the tests and lets be honest it's a little bit dirty:)
```
this.async(function () {
    this.$.dropdownItems.render();
})
```
I couldn't find a way how "fix" the test.

randyaskin commented:
I actually think it's fine if we remove the default value for sortMode, that way it will maintain whatever order the dev passes them in by default, but sort by key or val if either of those is specified - that's pretty much how I would expect it to work.

PiotrekPaprocki commented:
Awesome:) I'm glad I could help.

randyaskin commented:
Dang. Just when I thought we were in the clear. 😛 
I may have to take out some of the sorting, and displaying the list of selected items. Or at least make them configurable. It seems like a useful enhancement in some particular use cases, but it made things really confusing in our data-table, which makes use of the dropdown:

<img width="729" alt="screen shot 2017-03-20 at 4 21 47 pm" src="https://cloud.githubusercontent.com/assets/2130613/24126808/5e11d442-0d8d-11e7-907f-30e240e2bbca.png">

In this case, it really does make sense to keep the "Show/Hide Column" text in the control, cuz that's what it's being used for. And moving the hidden columns to the bottom could be confusing because the columns actually stay in the same order when they're hidden and then shown again.

PiotrekPaprocki commented:
oh yes, I didn't know about such use cases. I don't even use disabled items, but I thought I shouldn't left them be when sorting, unfortunately I was wrong ;p.  Now it's much more configurable so I think everyone are happy.
